base/special/exp10.jl

```diff
@@ -0,0 +1,131 @@
+#
+#                x = k*log10(2) + r,  |r| <= 0.5*log10(2).
+#
+# 2. Approximate exp10(r) by a polynomial on the interval [-0.5*log10(2), 0.5*log10(2)]:
+#
+#           exp10(x) = 1.0 + polynomial(x),
+#
+#    sup norm error within the interval specified, of the polynomial approximations:
+#    Float64 : [2.7245504724394698951963418186308011736517648015976e-18; 2.72455298957534767204462079073786522870226850867e-18]
+#    Float32 : [9.602647147784220587121706280697708280058577656746e-10; 9.6026560194009888671163788405817409356838220899077e-10]
+#
+# 3. Scale back: exp10(x) = 2^k * exp10(r)
+
+# log2(10)
+const LOG210 = 3.321928094887362347870319429489390175864831393024580612054756395815934776608624
+# log10(2)
+const LOG102 = 3.010299956639811952137388947244930267681898814621085413104274611271081892744238e-01
+# log(10)
+const LN10 = 2.302585092994045684017991454684364207601101488628772976033327900967572609677367
+
+# log10(2) into upper and lower bits
+LOG102U(::Type{Float64}) = 3.01025390625000000000e-1
+LOG102U(::Type{Float32}) = 3.00781250000000000000f-1
+
+LOG102L(::Type{Float64}) = 4.60503898119521373889e-6
+LOG102L(::Type{Float32}) = 2.48745663981195213739f-4
+
+# max and min arguments
+MAXEXP10(::Type{Float64}) = 3.08254715559916743851e2 # log 2^1023*(2-2^-52)
+MAXEXP10(::Type{Float32}) = 38.531839419103626f0     # log 2^127 *(2-2^-23)
+
+# one less than the min exponent since we can sqeeze a bit more from the exp10 function
+MINEXP10(::Type{Float64}) = -3.23607245338779784854769e2 # log10 2^-1075
+MINEXP10(::Type{Float32}) = -45.15449934959718f0         # log10 2^-150
+
+@inline exp10_kernel(x::Float64) =
+    @horner(x, 1.0,
+    2.30258509299404590109361379290930926799774169921875,
+    2.6509490552391992146397114993305876851081848144531,
+    2.03467859229323178027470930828712880611419677734375,
+    1.17125514891212478829629617393948137760162353515625,
+    0.53938292928868392106522833273629657924175262451172,
+    0.20699584873167015119932443667494226247072219848633,
+    6.8089348259156870502017966373387025669217109680176e-2,
+    1.9597690535095281527677713029333972372114658355713e-2,
+    5.015553121397981796436571499953060992993414402008e-3,
+    1.15474960721768829356725927226534622604958713054657e-3,
+    1.55440426715227567738830671828509366605430841445923e-4,
+    3.8731032432074128681303432086835414338565897196531e-5,
+    2.3804466459036747669197886523306806338950991630554e-3,
+    9.3881392238209649520573607528461934634833596646786e-5,
+    -2.64330486232183387018679354696359951049089431762695e-2)
+
+@inline exp10_kernel(x::Float32) =
+    @horner(x, 1.0f0,
+    2.302585124969482421875f0,
+    2.650949001312255859375f0,
+    2.0346698760986328125f0,
+    1.17125606536865234375f0,
+    0.5400512218475341796875f0,
+    0.20749187469482421875f0,
+    5.2789829671382904052734375f-2)
+
+@eval exp10_small_thres(::Type{Float64}) = $(2.0^-29)
+@eval exp10_small_thres(::Type{Float32}) = $(2.0f0^-13)
```

Comment:
Are you sure about this? How are you testing over all float32, can you post the script. For example I'm finding many numbers greater than 1 ulp. The max float32 ulp I have found is at least 1.158, yet where these occur are larger than the small threshold value, even the value you post (-0.000122065285f0) is larger than the threshold value

Here is my test script to double check

```julia
function countulp(::Type{T}, x::AbstractFloat, y::AbstractFloat) where T
    isnan(x) && isnan(y) && return zero(T)
    isnan(x) || isnan(y) && return T(1000)
    if isinf(T(x)) && isinf(T(y))
        return signbit(x) == signbit(y) ? zero(T) : T(1001)
    end
    if x == zero(x) && y == zero(y)
        return signbit(x) == signbit(y) ? zero(T) : T(1002)
    end
    if isfinite(x) && isfinite(y)
        return T(abs(x - y)/eps(T(y)))
    end
    return T(1003)
end

start32 = 0x00000000
end32 = 0xffffffff

function test_acc(f1,f2,::Type{T}, suint::U, euint::U, spacing::Integer=1000) where T where U<:Unsigned
    spacing_uint = convert(typeof(suint), spacing)
    vuint = suint
    x = reinterpret(T, vuint)
    ulp = zero(T)
    while vuint < euint
        v1 =f1(x)
        v2 = f2(big(x))
        ulp_ = countulp(T,v1,v2)
        ulp = max(ulp,ulp_)
        if ulp_ > T(1.0)
            @show x,ulp
        end
        vuint += spacing_uint
        x = reinterpret(T, vuint)
    end
    ulp
end

@time ulp,x = test_acc(Amal.exp10, Base.exp10, Float32, start32, end32)
```